docs/usage/WritingCustomMiddleware.md

```diff
@@ -0,0 +1,146 @@
+---
+id: writing-custom-middleware
+title: Writing Custom Middleware
+---
+
+# Writing Custom Middleware
+
+:::tip What You'll Learn
+
+- When to use custom middleware
+- Standard patterns for middleware
+- How to make sure that your middleware is compatible with other Redux projects
+
+:::
+
+Middleware in Redux can be mainly used to either
+
+- create side effects for actions, or to
+- modify or cancel actions.
+
+Most use cases fall into the first category: For example [Redux-Saga](https://github.com/redux-saga/redux-saga/), [redux-observable](https://github.com/redux-observable/redux-observable), and [RTK listener middleware](https://redux-toolkit.js.org/api/createListenerMiddleware) all create side effects that react to actions. These examples also show that this is a very common need: To be able to react to an action other than with a state change.
+
+The most obvious example for modifying actions is [Redux Thunk](https://github.com/reduxjs/redux-thunk), which transforms a function returning an action into an action by calling it. But it can also be used to e.g. enhance an action with information from the state or from an external input.
+
+## When to use custom middleware
+
+Most of the time, you won't actually need custom middleware. The most likely use case for middleware is side effects, and there is plenty of packages who nicely package side effects for Redux and have been in use long enough to get rid of the subtle problems you would run into when building this yourself. A good starting point is [RTK Query](https://redux-toolkit.js.org/rtk-query/overview) for managing server-side state and [RTK listener middleware](https://redux-toolkit.js.org/api/createListenerMiddleware) for other side effects.
+
+You might still want to use custom middleware in one of two cases:
+
+1. If you only have a single, very simple side effect, it might not be worth it to add a full additional framework. Just make sure that you switch to an existing framework once your application grows instead of growing your own custom solution.
+2. If you need to modify or cancel actions.
+
+## Standard patterns for middleware
+
+Middleware roughly works like this:
+
+```ts
+const middleware: Middleware = api => next => action => {
+  // Do something before the action hits the reducer and the next middlewares
+  const beforeState = api.getState()
+  beforeSideEffect(action, beforeState)
+  // Do something with the action before passing it along
+  const enhancedAction = enhance(action, beforeState, api)
+  let response: unknown
+  // Pass the action along, or cancel it by not passing it
+  if (!shouldCancelAction(enhancedAction, beforeState)) {
+    response = next(enhancedAction)
+  }
+  // Do something after the action hits the reducer
+  const afterState = api.getState()
+  afterSideEffect(action, afterState, api)
+
+  // Do something with the dispatch function return value
+  return enhanceDispatch(response, api)
```

Comment:
:question: not sure I understand what `enhanceDispatch` is here, or where it's coming from

Reply:
```suggestion
  return enhanceDispatchReturnValue(response, api)
```

Maybe this works better? Basically you can modify the return value of a dispatch function.

Reply:
I think there's still a couple things that bother me about this example.

One is that there's no indication what any of these functions do or where they're coming from. I think I understand what you're _trying_ to show - that you can run side effects before and after calling `next()`, etc.  I'm just not sure this example as written is the best way to explain that.

Second is that the `enhance` and `enhanceDispatch` functions in particular are very confusing even after re-reading this a few times.  I _think_ that `enhance` is trying to indicate that you can mutate the contents of the action.  I still don't know what `enhanceDispatchReturnValue` is supposed to do.

I think I'd much rather have this whole example explicitly show "real" examples of each kind of a thing, with comments right above each line explaining "here's an example of a side effect before calling `next(action)`", etc.  In other words, instead of calling a made-up function called `enhance(action)`, I'd rather see something like `action.meta.someField = 123`.  Similar for the side effect checks.

I don't immediately have good suggestions for what those examples ought to be.

Reply:
This example tries to cover all the points where you can "hook" into the lifecycle. E.g. `enhanceDispatchReturnValue` is used by redux-thunk to make dispatch awaitable. But most middlewares will only use one or maybe two of these points.
I like the idea of concrete examples. Maybe we could split it into the following:

1. `afterSideEffect` is very common, we could e.g. use RTK listener middleware for this.
2. `beforeSideEffect` is very similar to `afterSideEffect`, but I haven't seen this in use anywhere yet, as it doesn't allow access to the state after the action went through the reducer, so I would need help to find an example
3. `enhance` and `enhanceDispatch` can both be shown with Redux Thunk middleware
4. For stopping execution, I don't have a good example. Maybe something from RTK query? I don't have deep enough understanding of that code base for a good example, though.

Reply:
I'm okay with miniature examples, I'm just thinking it would be helpful if they showed actually _doing_ something instead of calling functions that don't actually exist in the code block :)

Reply:
I'll add a separate comment with a suggestion.